src/control/client/system.go

```diff
@@ -27,107 +27,122 @@ import (
 	"github.com/daos-stack/daos/src/control/system"
 )
 
 // SystemStopReq contains the inputs for the system stop command.
 type SystemStopReq struct {
-	Prep bool
-	Kill bool
+	Prep  bool
+	Kill  bool
+	Ranks []uint32
+}
+
+// SystemStopResp contains the request response.
+type SystemStopResp struct {
+	Results system.MemberResults
 }
 
 // SystemStop will perform a controlled shutdown of DAOS system and a list
 // of remaining system members on failure.
 //
 // Isolate protobuf encapsulation in client and don't expose to calling code.
-func (c *connList) SystemStop(req SystemStopReq) (system.MemberResults, error) {
+func (c *connList) SystemStop(req SystemStopReq) (*SystemStopResp, error) {
 	mc, err := chooseServiceLeader(c.controllers)
 	if err != nil {
 		return nil, err
 	}
 
 	rpcReq := &ctlpb.SystemStopReq{Prep: req.Prep, Kill: req.Kill}
 
-	c.log.Debug("Sending DAOS system shutdown request\n")
+	c.log.Debugf("DAOS system stop request: %s\n", rpcReq)
 
 	rpcResp, err := mc.getCtlClient().SystemStop(context.Background(), rpcReq)
 	if err != nil {
 		return nil, err
 	}
 
-	c.log.Debug("Received DAOS system shutdown response\n")
+	c.log.Debugf("DAOS system stop response: %s\n", rpcResp)
+
+	ssr := new(SystemStopResp)
+	if err := convert.Types(rpcResp, ssr); err != nil {
+		return nil, err
+	}
+	return ssr, nil
+}
+
+// SystemStartReq contains the inputs for the system start request.
+type SystemStartReq struct {
+	Ranks []uint32
+}
 
-	return proto.MemberResultsFromPB(c.log, rpcResp.Results), nil
+// SystemStartResp contains the request response.
+type SystemStartResp struct {
+	Results system.MemberResults // resulting from harness starts
 }
 
-// SystemStart will perform a restart after a controlled shutdown of DAOS system.
-func (c *connList) SystemStart() error {
+// SystemStart will perform a start after a controlled shutdown of DAOS system.
+func (c *connList) SystemStart(req SystemStartReq) (*SystemStartResp, error) {
 	mc, err := chooseServiceLeader(c.controllers)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	rpcReq := &ctlpb.SystemStartReq{}
 
-	c.log.Debugf("DAOS system restart request: %s\n", rpcReq)
+	c.log.Debugf("DAOS system start request: %s\n", rpcReq)
 
 	rpcResp, err := mc.getCtlClient().SystemStart(context.Background(), rpcReq)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	c.log.Debugf("DAOS system restart response: %s\n", rpcResp)
+	c.log.Debugf("DAOS system start response: %s\n", rpcResp)
 
-	return nil
+	ssr := new(SystemStartResp)
+	if err := convert.Types(rpcResp, ssr); err != nil {
+		return nil, err
+	}
+	return ssr, nil
 }
 
-// SystemQuery will return the list of members joined to DAOS system.
-//
-// Isolate protobuf encapsulation in client and don't expose to calling code.
-func (c *connList) SystemQuery() (system.Members, error) {
+// SystemQueryReq contains the inputs for the system query request.
+type SystemQueryReq struct {
+	Rank int32 // negative integer represents a NilRank
```

Comment:
This is an acceptable short-term workaround for the unfortunate dependency on netdetect, but it's not technically correct, and I would like to see a ticket created to address the technical debt. Using an int32 here instead of uint32 means that dmg cannot interact with ranks > 2147483647. This is exceedingly unlikely to cause any problems in the near future, but I can imagine rank schemes which use a method other than sequential assignment to encode information in the rank value.

Reply:
https://jira.hpdd.intel.com/browse/DAOS-4032